node/actors/bft/src/leader/state_machine.rs

```diff
@@ -29,11 +32,15 @@ pub(crate) struct StateMachine {
         validator::ViewNumber,
         HashMap<validator::PublicKey, validator::Signed<validator::ReplicaPrepare>>,
     >,
+    /// Prepare QC builders indexed by view number.
+    pub(crate) prepare_qc: BTreeMap<validator::ViewNumber, PrepareQCBuilder>,
     /// A cache of replica commit messages indexed by view number and validator.
     pub(crate) commit_message_cache: BTreeMap<
         validator::ViewNumber,
         HashMap<validator::PublicKey, validator::Signed<validator::ReplicaCommit>>,
     >,
+    /// Commit QC builders indexed by view number.
+    pub(crate) commit_qc: BTreeMap<validator::ViewNumber, CommitQCBuilder>,
 }
```

Comment:
Note that you no longer need prepare or commit message caches.

Reply:
Right, but it’s still used to check if we got a secondary message from a validator, which isn’t easily extractable from the certificate. 

Obviously a better solution can be made other than storing all messages, but I think we should settle on that only after the transition to non-blocking processing (BFT-406).